### DISCUSSION TITLE
Using ClosedXML with PowerShell

### DISCUSSION THREAD

**Discussion by dave-kimberley** [ClosedXML/ClosedXML#1970] [Q&A]
I have successfully used previous versions of ClosedXML with PowerShell by doing the following:

```
[Reflection.Assembly]::LoadFile((Join-Path $DLLPath 'ClosedXML.dll')) | Out-Null
[Reflection.Assembly]::LoadFile((Join-Path $DLLPath 'DocumentFormat.OpenXml.dll')) | Out-Null
[Reflection.Assembly]::LoadFile((Join-Path $DLLPath 'Fastmember.Signed.dll')) | Out-Null

$wb = New-Object ClosedXML.Excel.XLWorkbook
$ws = $wb.Worksheets.Add("ClosedXML - Testing")
$ws.Cell(1,1).Value = "ClosedXML - Testing"

$wb.SaveAs("ClosedXML-Testing.xlsx")
```

I have tried this using the latest version:

```
[Reflection.Assembly]::LoadFile((Join-Path $DLLPath 'ClosedXML.dll')) #| Out-Null
[Reflection.Assembly]::LoadFile((Join-Path $DLLPath 'DocumentFormat.OpenXml.dll')) #| Out-Null
[Reflection.Assembly]::LoadFile((Join-Path $DLLPath 'ExcelNumberFormat.dll')) #| Out-Null
[Reflection.Assembly]::LoadFile((Join-Path $DLLPath 'Irony.dll')) #| Out-Null
[Reflection.Assembly]::LoadFile((Join-Path $DLLPath 'SixLabors.Fonts.dll')) #| Out-Null
[Reflection.Assembly]::LoadFile((Join-Path $DLLPath 'System.Buffers.dll')) #| Out-Null
[Reflection.Assembly]::LoadFile((Join-Path $DLLPath 'System.IO.Packaging.dll')) #| Out-Null
[Reflection.Assembly]::LoadFile((Join-Path $DLLPath 'System.Memory.dll')) #| Out-Null
[Reflection.Assembly]::LoadFile((Join-Path $DLLPath 'System.Numerics.Vectors.dll')) #| Out-Null
[Reflection.Assembly]::LoadFile((Join-Path $DLLPath 'System.Runtime.CompilerServices.Unsafe.dll')) #| Out-Null
[Reflection.Assembly]::LoadFile((Join-Path $DLLPath 'System.ValueTuple.dll')) #| Out-Null
[Reflection.Assembly]::LoadFile((Join-Path $DLLPath 'XLParser.dll')) #| Out-Null

$wb = New-Object ClosedXML.Excel.XLWorkbook
$ws = $wb.Worksheets.Add("ClosedXML - Testing")
$ws.Cell(1,1).Value = "ClosedXML - Testing"

$wb.SaveAs((Join-Path $DLLPath "ClosedXML-Testing.xlsx"))
```

When running I get the following error:
```
Exception setting "Value": "Could not load file or assembly 'System.Runtime.CompilerServices.Unsafe, Version=4.0.4.1, Culture=neutral, 
PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified."
At C:\temp\PowerShell\Basic Test.ps1:28 char:1
+ $ws.Cell(1,1).Value = "ClosedXML - Testing"
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], ParentContainsErrorRecordException
    + FullyQualifiedErrorId : ExceptionWhenSetting
 ```

The DLLs all exist, I build a simple ClosedXML project in Visual Studio and copied them from it.

Any ideas on how I could get this to work?

Thanks

**Comment by jahav** [comment]
Sorry, I am not too familiar with powershell. From practical standpoint, the error reports version **4.0.4.1** (that is the assembly version), but lowest version for that library **4.0.6.0**. Note that these are assembly versions, not nuget package version, they differ.

4.0.4.1 seems to be a version from .NET Framework ~~4.6.1~~ 4.5.3 (at least Nuget says so, don't have it installed anymore). It is likely that something somewhere wants **4.0.4.1*, but gets a different version and thus just says "can't find anything"

When I get this, I generally just add assembly redirect, because 
> the assembly loader itself is basically made of unstable plutonium with a hair trigger coated in flesh eating bacteria behind a gate made of unobtanium above a moat of napalm filled with those jellyfish that kill you…that won’t ever really be fixed.
> [Binding Redirects](https://nickcraver.com/blog/2020/02/11/binding-redirects/)

In this case, the best thing I can offer is: google "powershell assembly redirect" and try something with that (also read the linked article, it has some helpful suggestions).

EDIT: Maybe try to use dll from the version 4.5.3: https://nuget.info/packages/System.Runtime.CompilerServices.Unsafe/4.5.3 (it's in the libr directory) - that version is 4.0.4.1 (it is displayed in the lower info, when you click on file) , but it is likely a game of whack a mole and probably won't work.
